blogs/2023/11/13/vscode-copilot-smarter.md

```diff
@@ -0,0 +1,175 @@
+
+## Agents
+
+Agents are domain experts that provide the large language model access to domain specific tools. With the help of the LLM, the agent may select a tool and define how to invoke it. One example of such an agent is `@workspace`. The `@workspace` agent knows about your workspace and can answer questions about it. Internally, the agent is powered by different tools:  GitHub's knowledge graph combined with semantic search, local code indexes, and the VS Code's language services. At the end of the day, an agent can do whatever it wants: it can implement the processing of the user query in a traditional way by forwarding it to a backend service, or the agent can fully use AI in the query processing.
+
+Agents can be contributed by the Client or the Service. At GitHub Universe, there was a demo of a service-side agent, a "Docs Agent" in the github.com Chat experience that knows how to search documentation found in repositories (coming soon to VS Code).
+
+Client-side agents can be contributed through traditional VS Code extensions. More on that in the [Extensibility](#extensibility) section, but let's take a look at the two agents that are available in VS Code today: `@workspace` and `@vscode`.
+
+### @workspace
+
+The `@workspace` agent knows how to gather context about the code in your workspace, can help you navigate it, find relevant classes, files, etc. Imagine you are in the [VS Code repository](https://github.com/microsoft/vscode) and you want to find out more about the service in charge of the current ICodeEditor; you can use the agent like this:
+
+![@workspace agent answering question about detecting running extensions](workspace-agent-example.png)
+
+We use natural language to ask the `workspace` agent "what service class do I use to get the current ICodeEditor". From there, the agent does the following to get just the right amount of context to send to the LLM:
+
+* The vscode repo is already indexed by the [GitHub Search Blackbird service]( https://github.blog/2023-02-06-the-technology-behind-githubs-new-code-search). The `@workspace` agent uses this index as a tool to tap into the repository knowledge graph. The `@workspace` agent runs a semantic search that returns relevant code snippets and meta data. GitHub search service has indexed the top 10K GitHub repositories, with plans to add more.
+
+* The next tool that the `@workspace` agent uses is the lexical text search over the local index to find extra code such as local, uncommitted changes, and the conversation history.
+
+* Then `@workspace` uses the final tool - VS Code"s language intelligence to add crucial details like function signatures, parameters, and even in-line documentation.
+
+All of these pieces of context are ranked, sliced, and summarized by the `@workspace`` agent and then sent off to the LLM to answer the question.
```

Comment:
There is a double backtick here `` after @workspace

Reply:
I hate that. I need to figure out the setting so that the editor doesn't keep adding a double ending backtick